File: Sources/Gantry/Services/MoonrakerClient.swift
import Foundation

/// Polls a Klipper printer's Moonraker API and reports status through MQTTClient.Event, so
/// PrinterStore treats Klipper and Bambu connections identically. REST polling keeps the first
/// version simple; a WebSocket subscription can replace the loop later.
final class MoonrakerClient: PrinterConnection, @unchecked Sendable {
    private let printer: SavedPrinter
    private let onEvent: @Sendable (MQTTClient.Event) -> Void
    private var task: Task<Void, Never>?
    private var telemetry = PrinterTelemetry()
    private var connectedReported = false
    private var disconnectReported = false

    // Creality CFS lives on the printer's own WebSocket, not Moonraker; fed in on the side.
    private var cfsTask: Task<Void, Never>?
    private var webSocket: URLSessionWebSocketTask?
    private let cfsLock = NSLock()
    private var cfsGroups: [FilamentGroup] = []
    private static let cfsRequest = "{\"method\":\"get\",\"params\":{\"boxsInfo\":1}}"

    private let objects: MoonrakerObjects

    init(printer: SavedPrinter, objects: MoonrakerObjects = .init(),
         onEvent: @escaping @Sendable (MQTTClient.Event) -> Void) {
        self.printer = printer
        self.objects = objects
        self.onEvent = onEvent
    }

    func start() {
        task = Task.detached(priority: .utility) { [weak self] in await self?.run() }
        cfsTask = Task.detached(priority: .utility) { [weak self] in await self?.runCFS() }
    }

    func stop() {
        task?.cancel()
        cfsTask?.cancel()
        webSocket?.cancel(with: .goingAway, reason: nil)
    }

    /// Sends a raw G-code line (used by the detail card's controls and automations). Klipper macros
    /// like PAUSE / RESUME / CANCEL_PRINT go through here too.
    func sendGcode(_ script: String) {
        Task.detached(priority: .utility) { [weak self] in
            guard let self,
                  var components = URLComponents(string: "\(self.baseURL)/printer/gcode/script") else { return }
            components.queryItems = [URLQueryItem(name: "script", value: script)]
            guard let url = components.url else { return }
            var request = URLRequest(url: url)
            request.httpMethod = "POST"
            request.timeoutInterval = 8
            if let key = self.printer.apiKey, !key.isEmpty { request.setValue(key, forHTTPHeaderField: "X-Api-Key") }
            _ = try? await URLSession.shared.data(for: request)
        }
    }

    private var baseURL: String { "http://\(printer.host):\(printer.port ?? 7125)" }

    private func run() async {
        guard let queryURL = await buildQueryURL() else {
            reportDisconnected("Nie znaleziono API Moonraker (port \(printer.port ?? 7125))")
            return
        }
        while !Task.isCancelled {
            do {
                let data = try await get(queryURL)
                if var updated = MoonrakerStatusParser.telemetry(from: data, previous: telemetry, objects: objects) {
                    // A Creality CFS overrides the (absent) Happy Hare gates with its own modules.
                    let cfs = currentCFSGroups()
                    if !cfs.isEmpty {
                        updated.filamentGroups = cfs
                        updated.amsSlots = cfs.flatMap(\.legacyAMSSlots)
                    }
                    telemetry = updated
                    if !connectedReported { connectedReported = true; onEvent(.connected) }
                    onEvent(.telemetry(updated))
                }
            } catch is CancellationError {
                return
            } catch {
                reportDisconnected(error.localizedDescription)
                return
            }
            try? await Task.sleep(for: .seconds(2))
        }
    }

    /// Query only objects the printer actually exposes, so Moonraker doesn't reject the request,
    /// and pick up any chamber temperature sensor and the MMU object if present.
    private func buildQueryURL() async -> URL? {
        var wanted = ["print_stats", "virtual_sdcard", "display_status", "mmu", "gcode_move",
                      objects.nozzle, objects.bed, objects.fan]
        if let chamber = objects.chamber { wanted.append(chamber) }
        wanted = Array(Set(wanted))
        if let listURL = URL(string: "\(baseURL)/printer/objects/list"),
           let data = try? await get(listURL),
           let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let objects = (root["result"] as? [String: Any])?["objects"] as? [String] {
            let available = Set(objects)
            wanted = wanted.filter { available.contains($0) }
            for object in objects where object.lowercased().contains("chamber")
                && (object.hasPrefix("temperature_sensor") || object.hasPrefix("heater_generic")) {
                wanted.append(object)
            }
        }
        guard !wanted.isEmpty else { return nil }
        var components = URLComponents(string: "\(baseURL)/printer/objects/query")
        components?.queryItems = wanted.map { URLQueryItem(name: $0, value: nil) }
        return components?.url
    }

    private func get(_ url: URL) async throws -> Data {
        var request = URLRequest(url: url)
        request.timeoutInterval = 8
        request.cachePolicy = .reloadIgnoringLocalCacheData
        if let key = printer.apiKey, !key.isEmpty { request.setValue(key, forHTTPHeaderField: "X-Api-Key") }
        let (data, response) = try await URLSession.shared.data(for: request)
        guard (response as? HTTPURLResponse)?.statusCode == 200 else { throw URLError(.badServerResponse) }
        return data
    }

    private func currentCFSGroups() -> [FilamentGroup] {
        cfsLock.lock(); defer { cfsLock.unlock() }
        return cfsGroups
    }

    private func setCFSGroups(_ groups: [FilamentGroup]) {
        cfsLock.lock(); cfsGroups = groups; cfsLock.unlock()
    }

    /// Best-effort Creality CFS reader. Connects to the printer's `ws://host:9999`, asks for
    /// `boxsInfo`, and maps any material boxes to AMS slots. Klipper printers without a CFS (Voron,
    /// stock Moonraker) simply never answer on that port, so this backs off and stays idle.
    private func runCFS() async {
        guard let url = URL(string: "ws://\(printer.host):9999") else { return }
        while !Task.isCancelled {
            let socket = URLSession.shared.webSocketTask(with: url)
            webSocket = socket
            socket.resume()

            let requester = Task { [weak self] in
                while !Task.isCancelled {
                    try? await self?.webSocket?.send(.string(MoonrakerClient.cfsRequest))
                    try? await Task.sleep(for: .seconds(5))
                }
            }
            do {
                while !Task.isCancelled {
                    let message = try await socket.receive()
                    let payload: Data?
                    switch message {
                    case .string(let text): payload = text.data(using: .utf8)
                    case .data(let data): payload = data
                    @unknown default: payload = nil
                    }
                    if let payload, let groups = MoonrakerStatusParser.parseCFSGroups(from: payload) {
                        setCFSGroups(groups)
                    }
                }
            } catch {
                // Connection refused/closed — expected on non-Creality printers.
            }
            requester.cancel()
            socket.cancel(with: .goingAway, reason: nil)
            webSocket = nil
            if Task.isCancelled { return }
            try? await Task.sleep(for: .seconds(30))   // gentle retry; don't hammer a closed port
        }
    }

    private func reportDisconnected(_ reason: String?) {
        guard !disconnectReported else { return }
        disconnectReported = true
        onEvent(.disconnected(reason))
    }
}

File: windows/Gantry.Windows/Services/MoonrakerClient.cs
using System.IO;
using System.Net.Http;
using System.Net.WebSockets;
using System.Text;
using System.Text.Json;
using Gantry.Models;

namespace Gantry.Services;

/// <summary>Polls a Klipper printer's Moonraker API and reports status through MqttEvent, so
/// PrinterStore treats Klipper and Bambu connections identically. Mirrors the macOS client.</summary>
public sealed class MoonrakerClient : IPrinterConnection
{
    private static readonly HttpClient Http = new() { Timeout = TimeSpan.FromSeconds(8) };
    private static readonly byte[] CfsRequest = Encoding.UTF8.GetBytes("{\"method\":\"get\",\"params\":{\"boxsInfo\":1}}");

    private readonly SavedPrinter _printer;
    private readonly Action<MqttEvent> _onEvent;
    private readonly CancellationTokenSource _cts = new();
    private PrinterTelemetry _telemetry = new();
    private bool _connectedReported;
    private bool _disconnectReported;

    // Creality CFS lives on the printer's own WebSocket, not Moonraker; fed in on the side.
    private readonly object _cfsLock = new();
    private List<FilamentGroup> _cfsGroups = new();

    private readonly MoonrakerObjects _objects;

    public MoonrakerClient(SavedPrinter printer, Action<MqttEvent> onEvent, MoonrakerObjects? objects = null)
    {
        _printer = printer;
        _onEvent = onEvent;
        _objects = objects ?? new MoonrakerObjects();
    }

    public void Start()
    {
        _ = Task.Run(RunAsync);
        _ = Task.Run(CfsLoopAsync);
    }

    public void Stop()
    {
        try { _cts.Cancel(); } catch { }
    }

    private string BaseUrl => $"http://{_printer.Host}:{_printer.Port ?? 7125}";

    private async Task RunAsync()
    {
        var query = await BuildQueryUrlAsync();
        if (query is null)
        {
            ReportDisconnected(string.Format(AppSettings.T("Moonraker API not found (port {0})"), _printer.Port ?? 7125));
            return;
        }
        while (!_cts.IsCancellationRequested)
        {
            try
            {
                var data = await GetAsync(query);
                var updated = MoonrakerStatusParser.Telemetry(data, _telemetry, _objects);
                if (updated is not null)
                {
                    // A Creality CFS overrides the (absent) Happy Hare gates with its own modules.
                    var cfs = CurrentCfsGroups();
                    if (cfs.Count > 0)
                    {
                        updated.FilamentGroups = cfs;
                        updated.AmsSlots = cfs.SelectMany(g => g.LegacyAmsSlots()).ToList();
                    }
                    _telemetry = updated;
                    if (!_connectedReported) { _connectedReported = true; _onEvent(new MqttEvent { Type = MqttEventType.Connected }); }
                    _onEvent(new MqttEvent { Type = MqttEventType.Telemetry, Telemetry = updated });
                }
            }
            catch (OperationCanceledException) { return; }
            catch (Exception ex) { ReportDisconnected(ex.Message); return; }
            try { await Task.Delay(TimeSpan.FromSeconds(2), _cts.Token); }
            catch (OperationCanceledException) { return; }
        }
    }

    private async Task<string?> BuildQueryUrlAsync()
    {
        var wanted = new List<string> { "print_stats", "virtual_sdcard", "display_status", "mmu", "gcode_move", _objects.Nozzle, _objects.Bed, _objects.Fan };
        if (!string.IsNullOrEmpty(_objects.Chamber)) wanted.Add(_objects.Chamber!);
        wanted = wanted.Distinct().ToList();
        try
        {
            var listData = await GetAsync($"{BaseUrl}/printer/objects/list");
            using var doc = JsonDocument.Parse(listData);
            if (doc.RootElement.TryGetProperty("result", out var res)
                && res.TryGetProperty("objects", out var objects) && objects.ValueKind == JsonValueKind.Array)
            {
                var available = objects.EnumerateArray().Select(o => o.GetString() ?? "").ToHashSet();
                wanted = wanted.Where(available.Contains).ToList();
                foreach (var name in available)
                    if (name.ToLowerInvariant().Contains("chamber")
                        && (name.StartsWith("temperature_sensor") || name.StartsWith("heater_generic")))
                        wanted.Add(name);
                // Every fan the machine publishes, not just `fan`: auxiliary and chamber fans live under
                // `fan_generic`/`heater_fan`/`controller_fan`, and some vendor forks have no bare `fan`.
                foreach (var name in available)
                    if (name == "fan" || name.StartsWith("fan_generic ") || name.StartsWith("heater_fan ")
                        || name.StartsWith("controller_fan ") || name.StartsWith("temperature_fan "))
                        wanted.Add(name);
                wanted = wanted.Distinct().ToList();
            }
        }
        catch { /* fall back to the default object set */ }

        if (wanted.Count == 0) return null;
        var query = string.Join("&", wanted.Select(Uri.EscapeDataString));
        return $"{BaseUrl}/printer/objects/query?{query}";
    }

    private async Task<byte[]> GetAsync(string url)
    {
        using var request = new HttpRequestMessage(HttpMethod.Get, url);
        if (!string.IsNullOrEmpty(_printer.ApiKey)) request.Headers.Add("X-Api-Key", _printer.ApiKey);
        using var response = await Http.SendAsync(request, _cts.Token);
        response.EnsureSuccessStatusCode();
        return await response.Content.ReadAsByteArrayAsync(_cts.Token);
    }

    /// Sends a raw G-code line (used by the detail card's controls and automations). Klipper macros
    /// like PAUSE / RESUME / CANCEL_PRINT go through here too.
    public void SendGcode(string script)
    {
        _ = Task.Run(async () =>
        {
            try
            {
                var url = $"{BaseUrl}/printer/gcode/script?script={Uri.EscapeDataString(script)}";
                using var request = new HttpRequestMessage(HttpMethod.Post, url);
                if (!string.IsNullOrEmpty(_printer.ApiKey)) request.Headers.Add("X-Api-Key", _printer.ApiKey);
                using var response = await Http.SendAsync(request);
            }
            catch { }
        });
    }

    private List<FilamentGroup> CurrentCfsGroups()
    {
        lock (_cfsLock) return new List<FilamentGroup>(_cfsGroups);
    }

    private void SetCfsGroups(List<FilamentGroup> groups)
    {
        lock (_cfsLock) _cfsGroups = groups;
    }

    /// <summary>Best-effort Creality CFS reader. Connects to the printer's ws://host:9999, asks for
    /// <c>boxsInfo</c>, and maps any material boxes to AMS slots. Klipper printers without a CFS
    /// (Voron, stock Moonraker) never answer on that port, so this backs off and stays idle.</summary>
    private async Task CfsLoopAsync()
    {
        var uri = new Uri($"ws://{_printer.Host}:9999");
        while (!_cts.IsCancellationRequested)
        {
            var socket = new ClientWebSocket();
            Task? requester = null;
            try
            {
                await socket.ConnectAsync(uri, _cts.Token);
                await socket.SendAsync(CfsRequest, WebSocketMessageType.Text, true, _cts.Token);
                requester = RequestCfsPeriodicallyAsync(socket);

                var buffer = new byte[16384];
                using var stream = new MemoryStream();
                while (!_cts.IsCancellationRequested && socket.State == WebSocketState.Open)
                {
                    stream.SetLength(0);
                    ValueWebSocketReceiveResult result;
                    do
                    {
                        result = await socket.ReceiveAsync(buffer.AsMemory(), _cts.Token);
                        if (result.MessageType == WebSocketMessageType.Close) break;
                        stream.Write(buffer, 0, result.Count);
                    } while (!result.EndOfMessage);
                    if (result.MessageType == WebSocketMessageType.Close) break;

                    var groups = MoonrakerStatusParser.ParseCfsGroups(stream.ToArray());
                    if (groups is not null) SetCfsGroups(groups);
                }
            }
            catch (OperationCanceledException) { return; }
            catch { /* connection refused/closed — expected on non-Creality printers */ }
            finally
            {
                if (requester is not null) { try { await requester; } catch { } }
                socket.Dispose();
            }

            if (_cts.IsCancellationRequested) return;
            try { await Task.Delay(TimeSpan.FromSeconds(30), _cts.Token); }
            catch (OperationCanceledException) { return; }
        }
    }

    private async Task RequestCfsPeriodicallyAsync(ClientWebSocket socket)
    {
        try
        {
            while (!_cts.IsCancellationRequested && socket.State == WebSocketState.Open)
            {
                await Task.Delay(TimeSpan.FromSeconds(5), _cts.Token);
                if (socket.State == WebSocketState.Open)
                    await socket.SendAsync(CfsRequest, WebSocketMessageType.Text, true, _cts.Token);
            }
        }
        catch { /* socket closed or cancelled */ }
    }

    private void ReportDisconnected(string? reason)
    {
        if (_disconnectReported) return;
        _disconnectReported = true;
        _onEvent(new MqttEvent { Type = MqttEventType.Disconnected, Reason = reason });
    }
}
